Jupyterlab adds jupytext metadata, but then `jupytext --sync` removes it again

## Use case

My team's use case: we edit `.ipynb` notebooks in Jupyterlab. We use jupytext to create companion `.py` files for code review.

Historically, we've relied on everyone installing jupytext locally to update the `.py` files after notebook edits. Now, we'd like to add `jupytext --sync` via pre-commit to ensure the files are kept in sync.

## Issue

- With jupytext installed locally, edits to the notebook via Jupyterlab also automatically update the companion `.py` file. jupyterlab/jupytext also adds a `jupytext.text_representation` dict to the notebook's `metadata`.
- When `jupytext --sync` runs, it removes `metadata.jupytext.text_representation` from the notebook file again.
- This causes problems when running `jupytext` via pre-commit: it updates the .ipynb notebook file to remove the extra jupytext metadata and then fails because the index is outdated:
```shell
$ git commit -a -m 'updated notebook'
jupytext.................................................................Failed
- hook id: jupytext
- exit code: 2
- files were modified by this hook

[jupytext] Reading test_notebook.ipynb in format ipynb
[jupytext] Loading test_notebook.py
[jupytext] Updating test_notebook.ipynb
[jupytext] Error: the git index is outdated.
Please add the paired notebook with:
    git add test_notebook.ipynb
[jupytext] Updating the timestamp of test_notebook.py
[jupytext] Reading test_notebook.py in format py
[jupytext] Loading test_notebook.ipynb
[jupytext] Error: the git index is outdated.
Please add the paired notebook with:
    git add test_notebook.ipynb
[jupytext] Updating the timestamp of test_notebook.py
```

Rerunning `git commit -a` works, since we add jupytext's change to the git index. However, this isn't ideal because it's confusing, and it's annoying to have to do this every time we edit notebooks.


### Details

I'm wondering why jupytext is adding this metadata in one context (editing via Jupyterlab) but removing it in another (called directly with `jupytext --sync`). I believe this is the code that removes the metadata: https://github.com/mwouts/jupytext/blob/57d83aaf6f0a776e7e45178d31fb6daa358c41f9/jupytext/jupytext.py#L461 (call stack: cli.main -> cli.jupytext_single_file -> lazy_write -> jupytext.writes)

It also seems that this only happens when using `--sync`. If I use `--from ipynb --to py` instead, then the original notebook isn't written, which avoids this issue. I'm using this as a workaround for now.


## Versions

This is happening with recent versions:

```shell
$ python -V
Python 3.8.12
$ pip list | grep -i jup
jupyter-client       7.1.0
jupyter-core         4.9.1
jupyter-server       1.13.1
jupyterlab           3.2.5
jupyterlab-pygments  0.1.2
jupyterlab-server    2.10.2
jupytext             1.13.5
```

Indeed with the new version of the test at https://github.com/mwouts/jupytext/commit/8e2f3bb84fc11da0aa3771ffc54a89ef2dd29498 I can reproduce the issue that you documented above (thanks for pointing out at the exact pattern!)
```
        # modify the ipynb file in Jupyter
        # Reload the notebook
        nb = cm.get("test.ipynb")['content']
        nb.cells.append(new_markdown_cell("A new cell"))
        cm.save(dict(type="notebook", content=nb), "test.ipynb")
    
        # The text representation metadata is in the py file
        assert "text_representation" in tmpdir.join("test.py").read()
        # But not in the ipynb notebook
>       assert "text_representation" not in tmpdir.join("test.ipynb").read()
E       assert 'text_representation' not in ('{\n'\n ' "cells": [\n'\n '  {\n'\n '   "cell_type": "markdown",\n'\n '   "id": "cell-1",\n'\n '   "metadata": {},\n'\n '   "source": [\n'\n '    "A short notebook"\n'\n '   ]\n'\n '  },\n'\n '  {\n'\n '   "cell_type": "markdown",\n'\n '   "id": "cell-3",\n'\n '   "metadata": {},\n'\n '   "source": [\n'\n '    "A new cell"\n'\n '   ]\n'\n '  }\n'\n ' ],\n'\n ' "metadata": {\n'\n '  "jupytext": {\n'\n '   "text_representation": {\n'\n '    "extension": ".py",\n'\n '    "format_name": "percent",\n'\n '    "format_version": "1.3",\n'\n '    "jupytext_version": "1.13.5"\n'\n '   }\n'\n '  },\n'\n '  "kernelspec": {\n'\n '   "display_name": "Python 3",\n'\n '   "language": "python",\n'\n '   "name": "python_kernel"\n'\n '  }\n'\n ' },\n'\n ' "nbformat": 4,\n'\n ' "nbformat_minor": 5\n'\n '}\n')
E         'text_representation' is contained here:
```

--- a/LICENSE
+++ b/LICENSE
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2018-2021 Marc Wouts
+Copyright (c) 2018-2022 Marc Wouts
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal

--- a/docs/CHANGELOG.md
+++ b/docs/CHANGELOG.md
@@ -1,11 +1,14 @@
 Jupytext ChangeLog
 ==================
 
-1.13.5 (2021-12-27)
+1.13.6 (2022-01-11)
 -------------------
 
 **Fixed**
 - The `text_representation` metadata of text notebooks is filtered from `.ipynb` files both in `jupytext.write` and in the contents manager for Jupyter ([#900](https://github.com/mwouts/jupytext/issues/900))
+
+**Changed**
+- Jupytext will not issue a warning when a format suffix starting with '.', '-' or '_' is passed to the `--to` option ([#901](https://github.com/mwouts/jupytext/issues/901))
 
 
 1.13.5 (2021-12-27)

--- a/docs/CHANGELOG.md
+++ b/docs/CHANGELOG.md
@@ -5,6 +5,13 @@ Jupytext ChangeLog
 -------------------
 
 **Fixed**
+- The `text_representation` metadata of text notebooks is filtered from `.ipynb` files both in `jupytext.write` and in the contents manager for Jupyter ([#900](https://github.com/mwouts/jupytext/issues/900))
+
+
+1.13.5 (2021-12-27)
+-------------------
+
+**Fixed**
 - Jupytext will not open a text notebook that is not UTF-8 ([#896](https://github.com/mwouts/jupytext/issues/896))
 
 

--- a/jupytext/cli.py
+++ b/jupytext/cli.py
@@ -452,7 +452,8 @@ def jupytext(args=None):
         not args.output
         and args.output_format
         and "." in args.output_format
-        and not args.output_format.startswith(".")
+        # a suffix is expected to start with one of these characters #901
+        and not args.output_format.startswith((".", "-", "_"))
         and "//" not in args.output_format
     ):
 
@@ -461,7 +462,7 @@ def jupytext(args=None):
 
         warnings.formatwarning = single_line
         warnings.warn(
-            "You have passed a file name to the '--to' option, "
+            "You might have passed a file name to the '--to' option, "
             "when a format description was expected. Maybe you want to use the '-o' option instead?"
         )
 

--- a/jupytext/contentsmanager.py
+++ b/jupytext/contentsmanager.py
@@ -31,6 +31,7 @@ from .formats import (
     short_form_multiple_formats,
     short_form_one_format,
 )
+from .jupytext import drop_text_representation_metadata
 from .kernels import set_kernelspec_from_language
 from .paired_paths import (
     InconsistentPath,
@@ -140,7 +141,15 @@ def build_jupytext_contents_manager_class(base_contents_manager_class):
 
                     self.create_prefix_dir(path, fmt)
                     if fmt["extension"] == ".ipynb":
-                        return self.super.save(model, path)
+                        return self.super.save(
+                            dict(
+                                type="notebook",
+                                content=drop_text_representation_metadata(
+                                    model["content"]
+                                ),
+                            ),
+                            path,
+                        )
 
                     if (
                         model["content"]["metadata"]

--- a/jupytext/contentsmanager.py
+++ b/jupytext/contentsmanager.py
@@ -14,8 +14,6 @@ try:
 except ImportError:
     pass
 
-import jupytext
-
 from .config import (
     JUPYTEXT_CONFIG_FILES,
     PYPROJECT_FILE,
@@ -31,7 +29,7 @@ from .formats import (
     short_form_multiple_formats,
     short_form_one_format,
 )
-from .jupytext import drop_text_representation_metadata
+from .jupytext import drop_text_representation_metadata, reads, writes
 from .kernels import set_kernelspec_from_language
 from .paired_paths import (
     InconsistentPath,
@@ -167,7 +165,7 @@ def build_jupytext_contents_manager_class(base_contents_manager_class):
                     text_model = dict(
                         type="file",
                         format="text",
-                        content=jupytext.writes(
+                        content=writes(
                             nbformat.from_dict(model["content"]), fmt=fmt, config=config
                         ),
                     )
@@ -220,7 +218,7 @@ def build_jupytext_contents_manager_class(base_contents_manager_class):
                     model["format"] = "json"
                     model["mimetype"] = None
                     try:
-                        model["content"] = jupytext.reads(
+                        model["content"] = reads(
                             model["content"], fmt=fmt, config=config
                         )
                     except Exception as err:
@@ -303,7 +301,7 @@ def build_jupytext_contents_manager_class(base_contents_manager_class):
                 text = self.super.get(
                     alt_path, content=True, type="file", format=format
                 )["content"]
-                return jupytext.reads(text, fmt=alt_fmt, config=config)
+                return reads(text, fmt=alt_fmt, config=config)
 
             inputs, outputs = latest_inputs_and_outputs(
                 path, fmt, formats, get_timestamp, contents_manager_mode=True

--- a/jupytext/jupytext.py
+++ b/jupytext/jupytext.py
@@ -454,20 +454,9 @@ def writes(notebook, fmt, version=nbformat.NO_CONVERT, config=None, **kwargs):
     ext = fmt["extension"]
     format_name = fmt.get("format_name")
 
-    jupytext_metadata = metadata.get("jupytext", {})
-
     if ext == ".ipynb":
-        # Remove jupytext section if empty
-        jupytext_metadata.pop("text_representation", {})
-        if not jupytext_metadata:
-            metadata.pop("jupytext", {})
         return nbformat.writes(
-            NotebookNode(
-                nbformat=notebook.nbformat,
-                nbformat_minor=notebook.nbformat_minor,
-                metadata=metadata,
-                cells=notebook.cells,
-            ),
+            drop_text_representation_metadata(notebook, metadata),
             version,
             **kwargs,
         )
@@ -481,6 +470,27 @@ def writes(notebook, fmt, version=nbformat.NO_CONVERT, config=None, **kwargs):
 
     writer = TextNotebookConverter(fmt, config)
     return writer.writes(notebook, metadata)
+
+
+def drop_text_representation_metadata(notebook, metadata=None):
+    """When the notebook is saved to an ipynb file, we drop the text_representation metadata"""
+    if metadata is None:
+        # Make a copy to avoid modification by reference
+        metadata = deepcopy(notebook["metadata"])
+
+    jupytext_metadata = metadata.get("jupytext", {})
+    jupytext_metadata.pop("text_representation", {})
+
+    # Remove the jupytext section if empty
+    if not jupytext_metadata:
+        metadata.pop("jupytext", {})
+
+    return NotebookNode(
+        nbformat=notebook["nbformat"],
+        nbformat_minor=notebook["nbformat_minor"],
+        metadata=metadata,
+        cells=notebook["cells"],
+    )
 
 
 def write(nb, fp, version=nbformat.NO_CONVERT, fmt=None, config=None, **kwargs):

--- a/jupytext/version.py
+++ b/jupytext/version.py
@@ -1,3 +1,3 @@
 """Jupytext's version number"""
 
-__version__ = "1.13.5"
+__version__ = "1.13.6-dev"

--- a/jupytext/version.py
+++ b/jupytext/version.py
@@ -1,3 +1,3 @@
 """Jupytext's version number"""
 
-__version__ = "1.13.6-dev"
+__version__ = "1.13.6"

--- a/tests/test_cm_config.py
+++ b/tests/test_cm_config.py
@@ -8,6 +8,7 @@ from nbformat.v4.nbbase import new_code_cell, new_markdown_cell, new_notebook
 from tornado.web import HTTPError
 
 import jupytext
+from jupytext import TextFileContentsManager
 from jupytext.compare import compare_cells
 
 from .utils import notebook_model
@@ -192,3 +193,32 @@ def test_metadata_filter_from_config_has_precedence_over_notebook_metadata(
 
     py = tmpdir.join("test.py").read()
     assert "notebook_metadata_filter: all" in py
+
+
+def test_test_no_text_representation_metadata_in_ipynb_900(
+    tmpdir,
+    python_notebook,
+):
+    tmpdir.join(".jupytext.toml").write('formats = "ipynb,py:percent"\n')
+
+    # create a test notebook and save it in Jupyter
+    nb = python_notebook
+    cm = TextFileContentsManager()
+    cm.root_dir = str(tmpdir)
+    cm.save(dict(type="notebook", content=nb), "test.ipynb")
+
+    # Assert that "text_representation" is in the Jupytext metadata #900
+    assert "text_representation" in tmpdir.join("test.py").read()
+    # But not in the ipynb notebook
+    assert "text_representation" not in tmpdir.join("test.ipynb").read()
+
+    # modify the ipynb file in Jupyter
+    # Reload the notebook
+    nb = cm.get("test.ipynb")["content"]
+    nb.cells.append(new_markdown_cell("A new cell"))
+    cm.save(dict(type="notebook", content=nb), "test.ipynb")
+
+    # The text representation metadata is in the py file
+    assert "text_representation" in tmpdir.join("test.py").read()
+    # But not in the ipynb notebook
+    assert "text_representation" not in tmpdir.join("test.ipynb").read()

--- a/tests/test_pre_commit_1_sync_with_config.py
+++ b/tests/test_pre_commit_1_sync_with_config.py
@@ -1,0 +1,110 @@
+import pytest
+from git.exc import HookExecutionError
+from nbformat.v4.nbbase import new_markdown_cell
+from pre_commit.main import main as pre_commit
+
+from jupytext import TextFileContentsManager, read, write
+
+from .utils import (
+    skip_pre_commit_tests_on_windows,
+    skip_pre_commit_tests_when_jupytext_folder_is_not_a_git_repo,
+)
+
+
+@skip_pre_commit_tests_on_windows
+@skip_pre_commit_tests_when_jupytext_folder_is_not_a_git_repo
+def test_pre_commit_hook_sync_with_config(
+    tmpdir,
+    cwd_tmpdir,
+    tmp_repo,
+    jupytext_repo_root,
+    jupytext_repo_rev,
+    python_notebook,
+):
+    pre_commit_config_yaml = f"""
+repos:
+- repo: {jupytext_repo_root}
+  rev: {jupytext_repo_rev}
+  hooks:
+  - id: jupytext
+    args: [--sync]
+"""
+    tmpdir.join(".pre-commit-config.yaml").write(pre_commit_config_yaml)
+
+    tmp_repo.git.add(".pre-commit-config.yaml")
+    pre_commit(["install", "--install-hooks", "-f"])
+
+    tmpdir.join(".jupytext.toml").write('formats = "ipynb,py:percent"\n')
+
+    # create a test notebook and save it in Jupyter
+    nb = python_notebook
+    cm = TextFileContentsManager()
+    cm.root_dir = str(tmpdir)
+    cm.save(dict(type="notebook", content=nb), "test.ipynb")
+
+    # Assert that "text_representation" is in the Jupytext metadata #900
+    assert "text_representation" in tmpdir.join("test.py").read()
+    # But not in the ipynb notebook
+    assert "text_representation" not in tmpdir.join("test.ipynb").read()
+
+    # try to commit it, should fail as the py version hasn't been added
+    tmp_repo.git.add("test.ipynb")
+    with pytest.raises(
+        HookExecutionError,
+        match="git add test.py",
+    ):
+        tmp_repo.index.commit("failing")
+
+    # add the py file, now the commit will succeed
+    tmp_repo.git.add("test.py")
+    tmp_repo.index.commit("passing")
+    assert "test.ipynb" in tmp_repo.tree()
+    assert "test.py" in tmp_repo.tree()
+
+    # The text representation metadata is still in the py file
+    assert "text_representation" in tmpdir.join("test.py").read()
+    # But not in the ipynb notebook
+    assert "text_representation" not in tmpdir.join("test.ipynb").read()
+
+    # modify the ipynb file in another editor
+    nb = read("test.ipynb")
+    nb.cells.append(new_markdown_cell("A new cell"))
+    write(nb, "test.ipynb")
+
+    tmp_repo.git.add("test.ipynb")
+
+    # We try to commit one more time, this updates the py file
+    with pytest.raises(
+        HookExecutionError,
+        match="files were modified by this hook",
+    ):
+        tmp_repo.index.commit("failing")
+
+    # the text file has been updated
+    assert "A new cell" in tmpdir.join("test.py").read()
+
+    # trying to commit should fail again because we forgot to add the .py version
+    with pytest.raises(HookExecutionError, match="git add test.py"):
+        tmp_repo.index.commit("still failing")
+
+    nb = read("test.ipynb")
+    assert len(nb.cells) == 2
+
+    # add the text file, now the commit will succeed
+    tmp_repo.git.add("test.py")
+    tmp_repo.index.commit("passing")
+
+    # modify the .py file
+    nb.cells.append(new_markdown_cell("A third cell"))
+    write(nb, "test.py", fmt="py:percent")
+    tmp_repo.git.add("test.py")
+
+    # the pre-commit hook will update the .ipynb file
+    with pytest.raises(HookExecutionError, match="git add test.ipynb"):
+        tmp_repo.index.commit("failing")
+
+    tmp_repo.git.add("test.ipynb")
+    tmp_repo.index.commit("passing")
+
+    nb = read("test.ipynb")
+    assert len(nb.cells) == 3

--- a/tests/test_pre_commit_1_sync_with_config.py
+++ b/tests/test_pre_commit_1_sync_with_config.py
@@ -3,7 +3,7 @@ from git.exc import HookExecutionError
 from nbformat.v4.nbbase import new_markdown_cell
 from pre_commit.main import main as pre_commit
 
-from jupytext import TextFileContentsManager, read, write
+from jupytext import TextFileContentsManager, read
 
 from .utils import (
     skip_pre_commit_tests_on_windows,
@@ -66,45 +66,23 @@ repos:
     # But not in the ipynb notebook
     assert "text_representation" not in tmpdir.join("test.ipynb").read()
 
-    # modify the ipynb file in another editor
-    nb = read("test.ipynb")
+    # modify the ipynb file in Jupyter
+    # Reload the notebook
+    nb = cm.get("test.ipynb")["content"]
     nb.cells.append(new_markdown_cell("A new cell"))
-    write(nb, "test.ipynb")
+    cm.save(dict(type="notebook", content=nb), "test.ipynb")
 
-    tmp_repo.git.add("test.ipynb")
+    # The text representation metadata is in the py file
+    assert "text_representation" in tmpdir.join("test.py").read()
+    # But not in the ipynb notebook
+    assert "text_representation" not in tmpdir.join("test.ipynb").read()
 
-    # We try to commit one more time, this updates the py file
-    with pytest.raises(
-        HookExecutionError,
-        match="files were modified by this hook",
-    ):
-        tmp_repo.index.commit("failing")
-
-    # the text file has been updated
+    # the text file has been updated (and the ipynb file as well)
     assert "A new cell" in tmpdir.join("test.py").read()
-
-    # trying to commit should fail again because we forgot to add the .py version
-    with pytest.raises(HookExecutionError, match="git add test.py"):
-        tmp_repo.index.commit("still failing")
-
     nb = read("test.ipynb")
     assert len(nb.cells) == 2
 
-    # add the text file, now the commit will succeed
-    tmp_repo.git.add("test.py")
-    tmp_repo.index.commit("passing")
-
-    # modify the .py file
-    nb.cells.append(new_markdown_cell("A third cell"))
-    write(nb, "test.py", fmt="py:percent")
-    tmp_repo.git.add("test.py")
-
-    # the pre-commit hook will update the .ipynb file
-    with pytest.raises(HookExecutionError, match="git add test.ipynb"):
-        tmp_repo.index.commit("failing")
-
+    # add both files, the commit will succeed
     tmp_repo.git.add("test.ipynb")
+    tmp_repo.git.add("test.py")
     tmp_repo.index.commit("passing")
-
-    nb = read("test.ipynb")
-    assert len(nb.cells) == 3

